shim/src/main.rs

```diff
@@ -445,42 +280,70 @@ fn get_version() -> &'static str {
         .0
 }
 
+/// Checks for `TURBO_BINARY_PATH` variable. If it is set,
+/// we do not do any inference, we simply run the command as
+/// the current binary. This is due to legacy behavior of `TURBO_BINARY_PATH`
+/// that lets users dynamically set the path of the turbo binary. Because
+/// inference involves finding a local turbo installation and executing that
+/// binary, these two features are fundamentally incompatible.
+fn is_turbo_binary_path_set() -> bool {
+    env::var("TURBO_BINARY_PATH").is_ok()
+}
+
 fn main() -> Result<()> {
-    let clap_args = Args::parse();
-    // --help doesn't work with ignore_errors in clap.
-    if clap_args.help {
-        let mut command = Args::command();
-        command.print_help()?;
-        process::exit(0);
-    }
-    // --version flag doesn't work with ignore_errors in clap, so we have to handle
-    // it manually
-    if clap_args.version {
-        println!("{}", get_version());
-        process::exit(0);
-    }
+    let mut clap_args = Args::new()?;
 
     let current_dir = if let Some(cwd) = &clap_args.cwd {
         fs::canonicalize::<PathBuf>(cwd.into())?
     } else {
         env::current_dir()?
     };
 
-    let args: Vec<_> = env::args().skip(1).collect();
-    if args.is_empty() {
-        process::exit(1);
+    clap_args.cwd = Some(current_dir.to_string_lossy().to_string());
+    if clap_args.test_run {
+        println!("{:#?}", clap_args);
+    }
+    // If there is no command, we set the command to `Command::Run` with
+    // `self.parsed_args.run_args` as arguments.
+    if clap_args.command.is_none() {
+        if let Some(run_args) = take(&mut clap_args.run_args) {
+            clap_args.command = Some(Command::Run(run_args));
+        } else {
+            return Err(anyhow!("No command specified"));
+        }
     }
 
-    let turbo_state = TurboState {
+    let mut turbo_state = TurboState {
+        repo_state: None,
         parsed_args: clap_args,
-        raw_args: env::args().skip(1).collect(),
     };
 
-    let exit_code = match turbo_state.run_correct_turbo(current_dir) {
-        Ok(exit_code) => exit_code,
-        Err(e) => {
-            eprintln!("failed {:?}", e);
-            2
+    // We run this *before* doing any inference because login/logout/link/unlink
+    // should work regardless of whether or not we're in a monorepo.
+    let exit_code = match turbo_state.parsed_args.command {
+        Some(Command::Login { .. })
+        | Some(Command::Link { .. })
+        | Some(Command::Logout { .. })
+        | Some(Command::Unlink { .. }) => turbo_state.native_run()?,
+        Some(Command::Completion { shell }) => {
+            generate(shell, &mut Args::command(), "turbo", &mut io::stdout());
+
+            0
+        }
+        _ => {
+            if is_turbo_binary_path_set() {
+                let repo_state = RepoState::infer(&current_dir)?;
+                turbo_state.repo_state = Some(repo_state);
+                turbo_state.native_run()?
```

Comment:
A comment explaining how `TURBO_BIN_PATH` is used during install and why us just checking if it's set is enough for us to know not to look for another binary. (I had to a knee jerk "this is wrong" reaction to this branch until I worked through the entire lifecycle in my head)

Reply:
I completely missed the comment 🤦